hyperbahn/advertise_test.go

```diff
@@ -261,48 +261,56 @@ func TestRetryTemporaryFailure(t *testing.T) {
 
 func TestRetryFailure(t *testing.T) {
 	runRetryTest(t, func(r *retryTest) {
-		r.mock.On("On", SendAdvertise).Return().
-			Times(1 /* initial */ + maxAdvertiseFailures /* fail */)
 		r.mock.On("On", Advertised).Return().Once()
+		r.mock.On("On", SendAdvertise).Return().Times(1 + (maxAdvertiseFailures * 2))
+
+		// Since FailStrategyIgnore will keep retrying forever, we need to
+		// signal when we're done testing.
+		doneTesting := make(chan struct{}, 1)
+		// For all but the last failure, we just need to assert that the
+		// OnError handler was called. The callback for the last failure needs
+		// to signal that we're done testing and close the Hyperbahn client (so
+		// that the Advertise call returns).
+		expectationCalled := 0
+		r.mock.On("OnError", ErrAdvertiseFailed{true, advertiseErr}).Return(nil).Times(maxAdvertiseFailures * 2).Run(func(_ mock.Arguments) {
+			expectationCalled++
+			if expectationCalled >= maxAdvertiseFailures*2 {
+				doneTesting <- struct{}{}
```

Comment:
I'll update to close this channel, since that's a good safety practice regardless.

However, but this isn't the channel that was being closed twice. That was expected behavior: the Hyperbahn client is closed once in the meat of the test (so AdvertiseLoop returns even though we never successfully advertise), and once in the shared test teardown code.

If we don't want to alter the Hyperbahn client's Close method so it's safe to call twice, I can tweak the test code to avoid double-closing the client. I'm a bit reluctant to make the testing code even more complex, though.